File: src/main/java/com/red/flink/source/PostgreSqlSource.java
package com.red.flink.source;

import org.apache.flink.api.common.ExecutionConfig;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
import org.apache.flink.types.Row;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.sql.*;
import java.util.Map;

/**
 * <b>自定义的Flink MySQL Source</b><br>
 *
 * <p>[详细描述]</p>
 * <p>
 * Date: 2022/7/6 17:30<br><br>
 *
 * @author 31528
 * @version 1.0
 */
public class PostgreSqlSource extends  RichParallelSourceFunction<Row>   {
    private PreparedStatement ps;
    private Connection connection;
    private Logger logger=LoggerFactory.getLogger(PostgreSqlSource.class);
    private String dbSource;
    private String tableName;

   public PostgreSqlSource(){
       super();
   }
   public PostgreSqlSource(String dbSource, String tableName){
       this.dbSource=dbSource;
       this.tableName=tableName;
   }


    /**
     * Reader JDBC Config
     * Create a connection
     * @param parameters
     * @throws Exception
     */
    @Override
    public void open(Configuration parameters) throws Exception {

        logger.info(String.format("============Class:%s Begin open method ======================",this.getClass().getName()));
        super.open(parameters);
        //Getting Global Configuration
        ExecutionConfig.GlobalJobParameters globalJobParameters =
                getRuntimeContext().getExecutionConfig().getGlobalJobParameters();
        Configuration globalConf=(Configuration) globalJobParameters;
        logger.info(String.format("Class=%s Configuration=%s", PostgreSqlSource.class.getName(),globalConf.toString()));

        //Reader JDBC Config
        Map<String, String> parametersMap = globalConf.toMap();
        String driver = parametersMap.get(String.format("jdbc.%s.driver", this.dbSource));
        String url = parametersMap.get(String.format("jdbc.%s.url", this.dbSource));
        String user = parametersMap.get(String.format("jdbc.%s.user", this.dbSource));
        String password = parametersMap.get(String.format("jdbc.%s.password", this.dbSource));

        logger.info(String.format("Create a connection"));
        Class.forName(driver);
        connection=DriverManager.getConnection(url,user,password);

        logger.info("Execute PrepareStatement SQL");
        ps=connection.
                prepareStatement(String.format("select * from %s",tableName));

    }

    @Override
    public void close() throws Exception {
        if(connection!=null){
            connection.close();
        }
        if(ps!=null){
            ps.close();
        }
    }

    /**
     * ResultSet convert  Row
     * @param sourceContext
     * @throws Exception
     */
    @Override
    public void run(SourceContext<Row> sourceContext) throws Exception {

       logger.info(String.format("============Class:%s Begin run method ======================",this.getClass().getName()));
        ResultSet resultSet = ps.executeQuery();
        int columnCount = resultSet.getMetaData().getColumnCount();

        // ResultSet convert to Row
        while (resultSet.next()){
            Row row = new Row(columnCount);
            for (int i = 1; i <=columnCount ; i++) {
                Object object = resultSet.getObject(i);
                //setField Begin pos 0
                row.setField(i-1,object);

            }
            //Collect Result Output
            sourceContext.collect(row);
        }

    }

    @Override
    public void cancel() {

    }

}

File: src/main/scala/com/red/flik/source/PostgreSqlSource.scala
package com.red.flik.source

import java.io.InputStream
import java.sql.{Connection, PreparedStatement, ResultSet}
import java.util
import java.util.Properties
import  com.esotericsoftware.kryo.serializers.ObjectField
import com.red.flik.util.ConnectionBuilder
import org.apache.flink.api.common.ExecutionConfig
import org.slf4j.{Logger, LoggerFactory}
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.source.{RichSourceFunction, SourceFunction}
import org.apache.flink.types.Row
/**
 * <b> PostgreSql Source</b><br>
 *
 * <p></p>
 *
 * Date: 2022/7/14 17:02<br><br>
 *
 * @author 31528
 * @version 1.0
 */
class PostgreSqlSource(dbSource:String,tableName:String) extends RichSourceFunction[Row]{

  private val logger: Logger = LoggerFactory.getLogger(this.getClass)
  var conn:Connection=null
  var ps:PreparedStatement=null

  /**
   * Create an Connection from  .properties file
   * @param parameters
   */
  override def open(parameters: Configuration): Unit = {
    logger.info(s"=================ClassName:${this.getClass.getName} Beginning Open method:Create an Connection=======================")
    super.open(parameters)
    val globalJobParameters: ExecutionConfig.GlobalJobParameters = getRuntimeContext.getExecutionConfig.getGlobalJobParameters
    val configMap: util.Map[String, String] = globalJobParameters.toMap
    val configPath:String= configMap get "configPath"
    logger.info(s"=================Load config file:$configPath=================================")
    val pathIn: InputStream = this.getClass.getResourceAsStream(configPath)
    val properties = new Properties()
        properties.load(pathIn)
    conn=ConnectionBuilder.getConnection(properties,dbSource)
    ps=conn.prepareStatement(s"select * from $tableName")
  }

  /**
   * Convert  PreparedStatement to Row
   * @param sourceContext
   */
  override def run(sourceContext: SourceFunction.SourceContext[Row]): Unit ={
    logger.info(s"============ClassName:${this.getClass.getName} Beginning run method: Convert  PreparedStatement to Row=============================")

    val rs: ResultSet = ps.executeQuery()
    val columnCount: Int = rs.getMetaData.getColumnCount
    logger.info("")

    while (rs.next()){
      var row = new Row(columnCount)
       for(i <- 1 to columnCount){
         // setField  pos 0-based
         row.setField(i-1,rs.getArray(i))
       }
      // Output result
      sourceContext.collect(row)

    }

  }

  override def close(): Unit = {
    logger.info(s"============ClassName:${this.getClass.getName} Beginning close method:close connection =============================")

    if (conn!=null) conn.close()
    if (ps!=null) ps.close()
  }

  override def cancel(): Unit = ???
}
